How many late days are you using for this assignment? 
0

In [23]:
# Imports and pip installations (if needed)
import pandas as pd


# Part 1: Load the dataset

In [24]:
# Load the given datasets
ds_1 = pd.read_csv('~/Downloads/chronic_kidney_disease_numerical.csv')
ds_2 = pd.read_csv('~/Downloads/chronic_kidney_disease_categorical.csv')
# Print the data
print(ds_1)
print(ds_2)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The first dataset (ds_1) seems to be storing numeric information of a patient's medical record. The features of the ds_1 are the age, blood pressure, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, and red blood cell count of a patient. There doesn't seem to be any labels in this dataset.

The second dataset (ds_2) also has information of a patient, but most of its features are categorical. The features of ds_2 are albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, and anemia. There's a label in this dataset, which is target. It's classifying each patient as having chronic kidney disease or not having chronic kidney disease.

The features in ds_1 are numerical whereas ds_2 includes both numerical and categorical features. Some examples of categorical features are appetite, red blood cells, and bacteria. Some examples of numerical features includes red blood cell count, white blood cell count, and hemoglobin.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [25]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = ds_1.duplicated(subset = ['unique_id'], keep = False).sum()
if num_duplicates > 0:
    print("number of duplicate rows in ds_1: ", num_duplicates)
# Drop these duplicate rows
unique_ds_1 = ds_1.drop_duplicates(subset = ['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
num_duplicates = ds_2.duplicated(subset = ['unique_id'], keep = False).sum()
if num_duplicates > 0:
    print("number of duplicate rows in ds_2: ", num_duplicates)
unique_ds_2 = ds_2.drop_duplicates(subset = ['unique_id'])

number of duplicate rows in ds_1:  59
number of duplicate rows in ds_2:  59


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: The unique ids are not present in both datasets because we know if we print out the two datasets after we dropped the duplicate rows, we can see that unique_ds_1 have dimensions 390 x 12 while unique_ds_2 have dimensions 400 x 14. Therefore, we know that there are unique ids in dataset 2 that are not present in dataset 1. That being said, in the combined table, we will see at least 10 rows with no values in the all the columns of dataset 1, such as age, blood pressure, etc.

In [26]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_ds = unique_ds_1.merge(unique_ds_2, on="unique_id", how="outer")
#Print the combined dataset
pd.set_option('display.max_columns', None)
print(merged_ds)


     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0   

        wbcc  rbcc   al   su     rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  no

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [27]:
# Calculate the percentage of rows that contain atleast one missing value
num_missing = merged_ds.isnull().any(axis = 1).sum()
num_rows = merged_ds.shape[0]
# Print %
print("percentage of rows that contain at least one missing value: "
      ,(num_missing/num_rows*100).round(2),"%")
# Drop these rows from the dataset
merged_ds = merged_ds.dropna()
# Print the Dataset
print(merged_ds)

percentage of rows that contain at least one missing value:  61.75 %
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su     rbc      

## Part 3.4: Sort the dataset according to the Labels

In [28]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_ds = merged_ds.sort_values(by = 'Target', ascending = True)
sorted_ds = sorted_ds.reset_index(drop = True)
# Print the dataset
print(sorted_ds)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [29]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_ds = pd.get_dummies(sorted_ds, columns = categorical_columns, dtype = int)
# Print the dataset
print(encoded_ds)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
0     3800.0   2.1  3.0  0.0            

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Yes, if we just keep a single column for a binary variable, this will be equivalent in the downstream modeling. However, if the categorical variable can take more than 2 values, that means we will have to assign certain integers to each value, such as 1 for ketchup, 2 for mayo, etc. This can then be misleading to the model as the dataset will only have 1 column with integers 1 to 5, but since the model doesn't know what each integer represent, it might think there's some meaningful relationship between these numbers, like how mayo (5) is greater than ketchup (1) which doesn't make sense.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [30]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

mean = encoded_ds[numerical_columns].mean()
std_dev = encoded_ds[numerical_columns].std()

no_outliers_ds = encoded_ds[((encoded_ds[numerical_columns]-mean).abs()<3*std_dev).all(axis=1)]
no_outliers_ds = no_outliers_ds.reset_index(drop = True)
# Print the dataset
print(no_outliers_ds)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
129     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0   
130     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0   
131     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0   
132     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0   
133     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
0     9200.0   3.2  2.0  0.0            

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [31]:
# Normalize the all Numerical Attributes in the dataset.
no_outliers_ds[numerical_columns] = no_outliers_ds[numerical_columns].apply(
    lambda x: (x - x.min()) / (x.max() - x.min())
)

# Print the dataset
print(no_outliers_ds)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
129     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
130     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
131     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
132     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
133     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv      wbcc      rbcc   al   su  rbc_abnormal  \
0    0.657143  0.172131  0.210

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column is not appropriate for modeling and predictions as it's just an unique identifier that we used at the beginning to merge the two datasets. However, it's not useful in predicting whether or not someone has chronic kidney disease so it shouldn't be treated as a feature which is why we will remove it from the dataset.

In [32]:
#Remove that column
cleaned_ds = no_outliers_ds.drop('unique_id', axis = 1)

# Print the dataset
print(cleaned_ds)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
..        ...   ...       ...       ...       ...       ...       ...   
129  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.257143   
130  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.400000   
131  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.371429   
132  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.171429   
133  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.514286   

         hemo       pcv      wbcc      rbcc   al   su  rbc_abnormal  \
0    0.172131  0.210526  0.395161  0.153846  2.0  0.

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [33]:
# Export the dataframe to a new csv file
cleaned_ds.to_csv('383_ckd_dataset.csv', index = False)

# Export the dataframe to a new json file
cleaned_ds.to_json('383_ckd_dataset.json', orient = 'index')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [34]:
print(cleaned_ds.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
5   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.028571   
6   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.542857   
7   0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.285714   
8   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286   
9   0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.514286   
10  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.571429   
11  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.714286   
12  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.57

Paste here: 
|    age    |    bp    |    bgr    |    bu    |    sc    |    sod    |    pot    |    hemo    |    pcv    |    wbcc    |    rbcc    |    al    |    su    | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|----------|-----------|----------|----------|-----------|-----------|------------|-----------|------------|------------|----------|----------|---------------|------------|-------------|-----------|----------------|--------------|---------------|-----------|--------|---------|-------|--------|--------|---------|-------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 0.743243  | 0.50     | 0.442060  | 0.901961 | 0.432099 | 0.500000  | 0.657143  | 0.172131   | 0.210526  | 0.395161   | 0.153846   | 2.0      | 0.0      | 1             | 0          | 1           | 0         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.770270  | 1.00     | 0.901288  | 0.163399 | 0.345679 | 0.766667  | 0.171429  | 0.606557   | 0.631579  | 0.443548   | 0.410256   | 2.0      | 2.0      | 0             | 1          | 0           | 1         | 1              | 0            | 0             | 1         | 0      | 1       | 1     | 0      | 0      | 1       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.905405  | 0.50     | 0.785408  | 0.862745 | 0.518519 | 0.600000  | 0.828571  | 0.401639   | 0.447368  | 0.233871   | 0.435897   | 2.0      | 0.0      | 1             | 0          | 1           | 0         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.202703  | 0.75     | 0.158798  | 0.196078 | 0.160494 | 0.166667  | 0.171429  | 0.221311   | 0.184211  | 0.653226   | 0.333333   | 4.0      | 0.0      | 0             | 1          | 1           | 0         | 0              | 1            | 0             | 1         | 1      | 0       | 1     | 0      | 1      | 0       | 1           | 0           | 1     | 0      | 0      | 1       | 1           | 0              |
| 0.783784  | 1.00     | 0.399142  | 0.287582 | 0.839506 | 0.666667  | 0.485714  | 0.188525   | 0.263158  | 0.258065   | 0.205128   | 4.0      | 2.0      | 1             | 0          | 1           | 0         | 1              | 0            | 0             | 1         | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0           | 0     | 1      | 1      | 0       | 1           | 0              |
| 0.729730  | 0.00     | 0.935622  | 0.169935 | 0.160494 | 0.333333  | 0.028571  | 0.188525   | 0.236842  | 0.879032   | 0.102564   | 3.0      | 1.0      | 0             | 1          | 1           | 0         | 0              | 1            | 1             | 0         | 0      | 1       | 1     | 0      | 0      | 1       | 0           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.675676  | 0.75     | 0.253219  | 0.633987 | 0.777778 | 0.366667  | 0.542857  | 0.286885   | 0.342105  | 0.169355   | 0.205128   | 2.0      | 0.0      | 1             | 0          | 1           | 0         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.851351  | 0.25     | 0.618026  | 0.562092 | 0.728395 | 0.000000  | 0.285714  | 0.311475   | 0.315789  | 0.580645   | 0.179487   | 4.0      | 3.0      | 0             | 1          | 1           | 0         | 0              | 1            | 0             | 1         | 0      | 1       | 0     | 1      | 0      | 1       | 1           | 0           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.540541  | 0.00     | 0.399142  | 0.535948 | 0.358025 | 0.700000  | 0.314286  | 0.344262   | 0.315789  | 0.830645   | 0.153846   | 1.0      | 0.0      | 0             | 1          | 0           | 1         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.756757  | 0.25     | 0.223176  | 0.209150 | 0.160494 | 0.533333  | 0.514286  | 0.573770   | 0.605263  | 0.290323   | 0.333333   | 3.0      | 0.0      | 0             | 1          | 1           | 0         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.662162  | 0.50     | 0.618026  | 0.411765 | 0.432099 | 0.566667  | 0.571429  | 0.434426   | 0.473684  | 0.250000   | 0.282051   | 3.0      | 1.0      | 0             | 1          | 1           | 0         | 0              | 1            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.783784  | 0.00     | 0.725322  | 0.313725 | 0.481481 | 0.566667  | 0.714286  | 0.319672   | 0.342105  | 0.258065   | 0.205128   | 4.0      | 1.0      | 1             | 0          | 1           | 0         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.878378  | 0.00     | 0.206009  | 0.751634 | 0.604938 | 0.533333  | 0.571429  | 0.475410   | 0.500000  | 0.879032   | 0.435897   | 4.0      | 0.0      | 0             | 1          | 0           | 1         | 1              | 0            | 1             | 0         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 0.878378  | 0.25     | 0.639485  | 0.470588 | 0.395062 | 0.433333  | 0.428571  | 0.393443   | 0.447368  | 0.104839   | 0.256410   | 3.0      | 0.0      | 0             | 1          | 1           | 0         | 0              | 1            | 0             | 1         | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 0.716216  | 0.50     | 1.000000  | 0.163399 | 0.111111 | 0.066667  | 0.171429  | 0.393443   | 0.500000  | 0.532258   | 0.435897   | 1.0      | 0.0      | 1             | 0          | 0           | 1         | 1              | 0            | 1             | 0         | 1      | 0       | 1     | 0      | 1      | 0       | 0           | 0           | 1     | 0      | 1      | 0       | 1           | 0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

In order to mitigate data conversion hallucinations from LLM conversions, you can maybe try feeding smaller chunks of data at a time so the LLM doesn't get overwhelmed by all the data it's receiving. Also, to make sure it's not hallucinating, check if you have the correct number of rows and columns. Also, if you know the values you are expecting, check that as well, like how we know the numerical columns should all have values between 0 and 1.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [35]:
# Code to rename all the columns in the dataset
rename_dict = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'CKD or Not CKD'
}

# Renaming the columns
cleaned_ds.rename(columns=rename_dict, inplace=True)

# Display the renamed DataFrame
print(cleaned_ds)

          Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0    0.743243            0.50              0.442060    0.901961   
1    0.770270            1.00              0.901288    0.163399   
2    0.905405            0.50              0.785408    0.862745   
3    0.202703            0.75              0.158798    0.196078   
4    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
129  0.189189            0.25              0.227468    0.222222   
130  0.216216            0.00              0.115880    0.052288   
131  0.513514            0.00              0.107296    0.235294   
132  0.702703            0.50              0.128755    0.261438   
133  0.689189            0.00              0.150215    0.254902   

     Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
0            0.432099  0.500000   0.657143    0.172131            0.210526   
1            0.345679  0.766667   0.171

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [36]:
# Converted SQL to Pandas code
result = cleaned_ds.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)

   Target_ckd  count
0           0    111
1           1     23


The code basically groups the rows according to its value in the 'Target_ckd' column and count how many rows has that value. More specifically, the result will only have two columns, which is outlined in the select clause of the SQL query, which is 'Target_ckd' and 'count'. Since we know 'Target_ckd' only has two values (0 and 1), we know there's only two groups, which corresponds to two rows in our result table. In addition, since 'count' represents the number of rows with that 'Target_ckd' value, the values in the 'count' should add up to 134, which is the number of rows in our cleaned dataset.